crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -332,56 +333,210 @@ fn build_new_replicated_loglet_configuration(
+        .unwrap_or_default();
 
-    let sequencer = preferred_sequencer
+    let &sequencer = preferred_sequencer
         .and_then(|node_id| {
             // map to a known alive node
             observed_cluster_state.alive_nodes.get(&node_id.id())
         })
         .or_else(|| {
             // we can place the sequencer on any alive node
             observed_cluster_state.alive_nodes.values().choose(&mut rng)
-        });
-
-    if log_servers.len() >= 3 {
-        let replication = ReplicationProperty::new(NonZeroU8::new(2).expect("to be valid"));
-        let mut nodeset = NodeSet::empty();
+        })?;
 
-        for node in &log_servers {
-            nodeset.insert(node.as_plain());
-        }
+    let selection = NodeSetSelector::new(nodes_config.as_ref(), observed_cluster_state).select(
+        strategy,
+        &replication,
+        &mut rng,
+        &preferred_nodes,
+    );
 
-        Some(ReplicatedLogletParams {
+    match selection {
+        Ok(nodeset) => Some(ReplicatedLogletParams {
             loglet_id,
-            sequencer: *sequencer.expect("one node must exist"),
+            sequencer,
             replication,
             nodeset,
             write_set: None,
-        })
-    } else if let Some(sequencer) = sequencer {
-        previous_configuration.cloned().map(|mut configuration| {
-            configuration.loglet_id = loglet_id;
-            configuration.sequencer = *sequencer;
-            configuration
-        })
-    } else {
-        None
+        }),
+        Err(NodeSelectionError::InsufficientWriteableNodes) => {
+            debug!(
+                ?loglet_id,
+                "Insufficient writeable nodes to place replicated loglet"
+            );
+            None
+        }
+    }
+}
+
+#[derive(
+    Debug, Clone, Eq, PartialEq, derive_more::Display, derive_more::Into, derive_more::IntoIterator,
+)]
+struct WritableNodeSet(NodeSet);
+
+impl WritableNodeSet {
+    /// Constructs a new nodeset consisting of only alive, read-write storage nodes.
+    pub fn from(cluster_state: &ObservedClusterState, nodes_config: &NodesConfiguration) -> Self {
+        Self(
+            cluster_state
+                .alive_nodes
+                .keys()
+                .copied()
+                .filter(|node_id| {
+                    nodes_config
+                        .get_log_server_storage_state(node_id)
+                        .can_write_to()
+                })
+                .collect(),
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// Nodeset selector for picking a set of storage nodes for a replicated loglet out of a
+/// broader pool of available nodes.
+///
+/// This selector can be reused once constructed to make multiple decision in a single scheduling iteration,
+/// if the node configuration and the replication settings are not changing.
+#[cfg(feature = "replicated-loglet")]
+#[derive(Clone)]
+struct NodeSetSelector<'a> {
```

Comment:
Feel free to move the `NodeSetSelector` to its own module to reduce the length of this file a bit.